priam/src/main/java/com/netflix/priam/config/PriamConfigurationPersister.java

```diff
@@ -0,0 +1,109 @@
+package com.netflix.priam.config;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.HashMap;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.util.MinimalPrettyPrinter;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.netflix.priam.scheduler.CronTimer;
+import com.netflix.priam.scheduler.Task;
+import com.netflix.priam.scheduler.TaskTimer;
+
+/**
+ * Task that persists structured and merged priam configuration to disk.
+ */
+@Singleton
+public class PriamConfigurationPersister extends Task
+{
+    public static final String NAME = "PriamConfigurationPersister";
+
+    private static final Logger logger = LoggerFactory.getLogger(PriamConfigurationPersister.class);
+
+    private final Path mergedConfigDirectory;
+    private final Path structuredPath;
+
+    private boolean initialized;
+    private boolean canWrite;
+
+    @Inject
+    public PriamConfigurationPersister(IConfiguration config) {
+        super(config);
+
+        mergedConfigDirectory = Paths.get(config.getMergedConfigurationDirectory());
+        structuredPath = Paths.get(config.getMergedConfigurationDirectory(), "structured.json");
+
+        initialized = false;
+        canWrite = false;
+    }
+
+    private synchronized void maybeSetupPaths() throws IOException
+    {
+        if (initialized)
+            return;
+
+        File directory = mergedConfigDirectory.toFile();
+        canWrite = (
+            directory.mkdirs() || (directory.exists() && directory.isDirectory())
+        );
+
+        if (canWrite)
+            Files.setPosixFilePermissions(mergedConfigDirectory, PosixFilePermissions.fromString("rwx------"));
+
+
+        initialized = true;
+        logger.info("Done setting up PriamConfigurationPersister, canWrite=" + canWrite);
+    }
+
+
+    @Override
+    public void execute() throws Exception
+    {
+        if (!initialized)
+            maybeSetupPaths();
+
+        Path structuredPathTmp = structuredPath.resolveSibling(structuredPath.getFileName() + ".tmp");
+        File output = structuredPathTmp.toFile();
+        output.createNewFile();
+        // The configuration might contain sensitive information, so ... don't let non Priam users read it
+        Files.setPosixFilePermissions(structuredPathTmp, PosixFilePermissions.fromString("rw-------"));
```

Comment:
I believe Files.createTempFile will do the work of opening a file with the desired permissions for you.

Reply:
Hm, I didn't want to use that in case there was a bug and I created thousands of forgotten tmp files (like the last time that happened in Priam) ... But sure it's shorter

I don't want to implicitly rely on a non public guarantee of createTempFile though (afaik it doesn't guarantee the permissions). I'm going to leave the chmod just in case.

Reply:
I ended up doing this just if the rename fails I delete it... hopefully there is no exception